### PR TITLE
Remove legacy DataFrame implementation

### PR DESCRIPTION
**Pull request by phofl**
- [ ] Closes #xxxx
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

TODOs:

- test cleanup (followup PR)

### CODE REVIEW

File: continuous_integration/environment-3.10.yaml [hunk 1]
```diff
@@ -76,5 +76,5 @@ dependencies:
   - jinja2
   - pip
   - pip:
-    - git+https://github.com/dask/distributed
-    - git+https://github.com/dask/dask-expr
+    - git+https://github.com/phofl/distributed@remove-legacy-implementation
```
Comment (hendrikmakait) [inline continuous_integration/environment-3.10.yaml:79]: TODO: Revert these changes

File: continuous_integration/scripts/test_imports.sh [hunk 2]
```diff
@@ -11,6 +11,9 @@ test_import () {
         # dask[distributed] depends on the latest version of distributed
         python -m pip install git+https://github.com/dask/distributed
     fi
+    if [[ $1 =~ "pandas dask-expr" ]]; then
```
Comment (hendrikmakait) [inline continuous_integration/scripts/test_imports.sh:14]: TODO: Revert these changes

File: continuous_integration/environment-3.11.yaml [hunk 3]
```diff
@@ -74,5 +74,5 @@ dependencies:
   - jinja2
   - pip
   - pip:
-    - git+https://github.com/dask/distributed
-    - git+https://github.com/dask/dask-expr
+    - git+https://github.com/phofl/distributed@remove-legacy-implementation
+    - git+https://github.com/phofl/dask-expr@legacy-conversion
```
Comment (hendrikmakait) [inline continuous_integration/environment-3.11.yaml]: TODO: Revert these changes

File: continuous_integration/environment-3.10.yaml [hunk 4]
```diff
@@ -76,5 +76,5 @@ dependencies:
...
+    - git+https://github.com/phofl/dask-expr@legacy-conversion
```
Comment (phofl) [inline continuous_integration/environment-3.10.yaml:80]:
```suggestion
    - git+https://github.com/dask/dask-expr
```

File: continuous_integration/environment-3.11.yaml (hunk 3 above)
Comment (phofl) [inline continuous_integration/environment-3.11.yaml:78]:
```suggestion
    - git+https://github.com/dask/dask-expr
```

File: continuous_integration/environment-3.12.yaml [hunk 5]
```diff
@@ -75,5 +75,5 @@ dependencies:
   - matplotlib  # visualize test
   - pip
   - pip:
-    - git+https://github.com/dask/distributed
-    - git+https://github.com/dask/dask-expr
+    - git+https://github.com/phofl/distributed@remove-legacy-implementation
+    - git+https://github.com/phofl/dask-expr@legacy-conversion
```
Comment (phofl) [inline continuous_integration/environment-3.12.yaml:79]:
```suggestion
    - git+https://github.com/dask/dask-expr
```

File: continuous_integration/environment-3.13.yaml [hunk 6]
```diff
@@ -72,6 +72,6 @@ dependencies:
   - jinja2
   - pip
   - pip:
-    - git+https://github.com/dask/distributed
+    - git+https://github.com/phofl/distributed@remove-legacy-implementation
     - mmh3
-    - git+https://github.com/dask/dask-expr
+    - git+https://github.com/phofl/dask-expr@legacy-conversion
```
Comment (phofl) [inline continuous_integration/environment-3.13.yaml:77]:
```suggestion
    - git+https://github.com/dask/dask-expr
```

File: continuous_integration/environment-mindeps-dataframe.yaml [hunk 7]
```diff
@@ -23,4 +23,4 @@ dependencies:
   - pytest-timeout
   - pytest-xdist
   - pip:
-    - git+https://github.com/dask/dask-expr
+    - git+https://github.com/phofl/dask-expr@legacy-conversion
```
Comment (phofl) [inline continuous_integration/environment-mindeps-dataframe.yaml:26]:
```suggestion
    - git+https://github.com/dask/dask-expr
```

File: continuous_integration/environment-mindeps-optional.yaml [hunk 8]
```diff
@@ -51,8 +51,8 @@ dependencies:
   - pip
   - pip:
       # optional dependencies pulled in by pip install dask[distributed]
-      - git+https://github.com/dask/distributed
-      - git+https://github.com/dask/dask-expr
+      - git+https://github.com/phofl/distributed@remove-legacy-implementation
+      - git+https://github.com/phofl/dask-expr@legacy-conversion
```
Comment (phofl) [inline continuous_integration/environment-mindeps-optional.yaml:55]:
```suggestion
    - git+https://github.com/dask/dask-expr
```

File: continuous_integration/scripts/test_imports.sh [hunk 9]
```diff
@@ -11,6 +11,9 @@ test_import () {
...
+        python -m pip install git+https://github.com/phofl/dask-expr@legacy-conversion -U
```
Comment (phofl) [inline continuous_integration/scripts/test_imports.sh:15]:
```suggestion
        python -m pip install git+https://github.com/dask/dask-expr -U
```

File: continuous_integration/environment-3.11.yaml [hunk 10]
```diff
@@ -74,5 +74,5 @@ dependencies:
   - jinja2
   - pip
   - pip:
-    - git+https://github.com/dask/distributed
+    - git+https://github.com/phofl/distributed@remove-legacy-implementation
```
Comment (hendrikmakait) [inline continuous_integration/environment-3.11.yaml:77]: TODO: Revert these changes

File: continuous_integration/environment-3.12.yaml [hunk 11]
```diff
@@ -75,5 +75,5 @@ dependencies:
   - matplotlib  # visualize test
   - pip
   - pip:
-    - git+https://github.com/dask/distributed
+    - git+https://github.com/phofl/distributed@remove-legacy-implementation
```
Comment (hendrikmakait) [inline continuous_integration/environment-3.12.yaml:78]: TODO: Revert these changes

File: dask/dataframe/__init__.py [hunk 12]
```diff
@@ -1,206 +1,94 @@
...
 
 
 def _dask_expr_enabled() -> bool:
     import dask
 
-    global DASK_EXPR_ENABLED
-
     use_dask_expr = dask.config.get("dataframe.query-planning")
+    if use_dask_expr is False:
+        raise NotImplementedError("The legacy implementation is no longer supported")
```
Comment (hendrikmakait) [inline dask/dataframe/__init__.py:11]: nit: `ValueError` might be more applicable here.

File: continuous_integration/environment-mindeps-optional.yaml [hunk 13]
```diff
@@ -51,8 +51,8 @@ dependencies:
   - pip
   - pip:
       # optional dependencies pulled in by pip install dask[distributed]
-      - git+https://github.com/dask/distributed
-      - git+https://github.com/dask/dask-expr
+      - git+https://github.com/phofl/distributed@remove-legacy-implementation
+    - git+https://github.com/dask/dask-expr
```
Comment (phofl) [inline continuous_integration/environment-mindeps-optional.yaml:55]:
```suggestion
      - git+https://github.com/dask/dask-expr
```

File: docs/requirements-docs.txt [hunk 14]
```diff
@@ -19,7 +19,7 @@ toolz
 cloudpickle>=1.5.0
 pandas>=2.0.0
 git+https://github.com/dask/distributed
-git+https://github.com/dask-contrib/dask-expr
+git+https://github.com/phofl/dask-expr@legacy-conversion
```
Comment (hendrikmakait) [inline docs/requirements-docs.txt]: TODO: Revert these changes

Comment (phofl) [inline docs/requirements-docs.txt:22]:
```suggestion
git+https://github.com/dask/dask-expr
```

**Review by hendrikmakait — APPROVED** [review]
Thanks, @phofl. LGM as long as CI is green. One nit, the rest are reminders to revert changes to the linked Github repos.

File: continuous_integration/environment-3.13.yaml [hunk 15]
```diff
@@ -72,6 +72,6 @@ dependencies:
   - jinja2
   - pip
   - pip:
-    - git+https://github.com/dask/distributed
+    - git+https://github.com/phofl/distributed@remove-legacy-implementation
```
Comment (hendrikmakait) [inline continuous_integration/environment-3.13.yaml:75]: TODO: Revert these changes

File: continuous_integration/environment-mindeps-optional.yaml [hunk 16]
```diff
@@ -51,7 +51,7 @@ dependencies:
   - pip
   - pip:
       # optional dependencies pulled in by pip install dask[distributed]
-      - git+https://github.com/dask/distributed
+      - git+https://github.com/phofl/distributed@remove-legacy-implementation
```
[the remark this answers is not in the document]
Comment (phofl) [reply continuous_integration/environment-mindeps-optional.yaml:54]:
```suggestion
      - git+https://github.com/dask/distributed
```

File: continuous_integration/environment-3.13.yaml (hunk 15 above)
Comment (phofl) [reply continuous_integration/environment-3.13.yaml:75]:
```suggestion
    - git+https://github.com/dask/distributed
```

File: continuous_integration/environment-3.12.yaml (hunk 11 above)
Comment (phofl) [reply continuous_integration/environment-3.12.yaml:78]:
```suggestion
    - git+https://github.com/dask/distributed
```

File: continuous_integration/environment-3.11.yaml (hunk 10 above)
Comment (phofl) [reply continuous_integration/environment-3.11.yaml:77]:
```suggestion
    - git+https://github.com/dask/distributed
```

File: continuous_integration/environment-3.10.yaml (hunk 1 above)
Comment (phofl) [reply continuous_integration/environment-3.10.yaml:79]:
```suggestion
    - git+https://github.com/dask/distributed
```

**Comment by rjzamora** [comment]
Wow, this is a very big PR! Hopefully it wont take too long for us to find/fix all the broken edges in rapids :)

Thanks for working on this @phofl - I know it's challenging to make a migration like this.

**Comment by phofl** [comment]
@rjzamora any chance you can provide a list of things you import from dask.dataframe.foo so that we can think if providing a proper downstream packages API makes sense? I suspect that you are also importing stuff from dask_expr, which will go away soon and thus require another update on your end? Putting all of this in some folder that we know that we shouldn't move in the future would make things easier

**Comment by rjzamora** [comment]
>any chance you can provide a list of things you import from dask.dataframe.foo so that we can think if providing a proper downstream packages API makes sense?

Yes, I can definitely put this together. Thanks for the suggestion!

>I suspect that you are also importing stuff from dask_expr, which will go away soon and thus require another update on your end? Putting all of this in some folder that we know that we shouldn't move in the future would make things easier

Yes, we definitely import (and patch) a lot of stuff from `dask_expr`. I was expecting RAPIDS would need to pin and revise after dask-expr was "mopved" into `dask/dask`, but it definitely makes sense for us to share the specific APIs we want to treat as "public".

**Comment by rjzamora** [comment]
Just following up... I compiled a rough list of relevant imports within Dask cuDF and Dask-CUDA:

<details >
<summary>Imports from Dask DataFrame</summary>
<br>

```
dd._dask_expr_enabled  # Was using dd.DASK_EXPR_ENABLED before 11606
dd.DataFrame
dd.Series
dd.concat
dd.from_delayed
dd.from_map
dd.from_pandas
dd.read_csv
dd.read_json
dd.read_orc
dd.read_parquet
dd.to_parquet

dd.backends.DataFrameBackendEntrypoint
dd.backends.PandasBackendEntrypoint

dd.core.is_dataframe_like
dd.core._concat

dd.dispatch.categorical_dtype_dispatch
dd.dispatch.concat
dd.dispatch.concat_dispatch
dd.backends.concat_pandas
dd.dispatch.from_pyarrow_table_dispatch
dd.dispatch.get_parallel_type
dd.dispatch.group_split_dispatch
dd.dispatch.grouper_dispatch
dd.dispatch.hash_object_dispatch
dd.dispatch.is_categorical_dtype
dd.dispatch.is_categorical_dtype_dispatch
dd.dispatch.make_meta
dd.dispatch.make_meta_dispatch
dd.dispatch.meta_nonempty
dd.dispatch.partd_encode_dispatch
dd.dispatch.pyarrow_schema_dispatch
dd.dispatch.to_pyarrow_table_dispatch
dd.dispatch.tolist_dispatch
dd.dispatch.union_categoricals_dispatch

dd.groupby.Aggregation

dd.io.csv.make_reader
dd.io.parquet.arrow.ArrowDatasetEngine
dd.io.parquet.arrow._filters_to_expression
dd.io.parquet.core.ParquetFunctionWrapper
dd.io.parquet.create_metadata_file
dd.io.utils._get_pyarrow_dtypes
dd.io.utils._is_local_fs

dd.shuffle.partitioning_index
dd.shuffle.shuffle_group

dd.utils.UNKNOWN_CATEGORIES
dd.utils._nonempty_scalar
dd.utils._scalar_from_dtype
dd.utils.assert_eq
dd.utils.is_dataframe_like
dd.utils.is_series_like
dd.utils.is_index_like
dd.utils.make_meta_obj
dd.utils.make_scalar
dd.utils.pyarrow_strings_enabled
```

</details>

I don't think any of these `dask.dataframe` imports are a real problem (though I could be missing something).

One problem caused by this PR is that we have been using `dd.DASK_EXPR_ENABLED` to check if query-planning is enabled in `dask.dataframe`. Now that `DASK_EXPR_ENABLED` is gone, we are stuck using `_dask_expr_enabled` until we decide to pin to the latest Dask version.

<details>
<summary>Imports from Dask Expressions</summary>
<br>

```
dx.from_dict
dx.get_collection_type
dx.new_collection
dx.DataFrame
dx.FrameBase
dx.Index
dx.Series

dx.io.csv.ReadCSV
dx.io.io.FusedIO
dx.io.io.FusedParquetIO
dx.io.parquet.FragmentWrapper
dx.io.parquet.ReadParquetFSSpec
dx.io.parquet.ReadParquetPyarrowFS

dx._cumulative.CumulativeBlockwise

dx._expr.Elemwise
dx._expr.Expr
dx._expr.RenameAxis
dx._expr.VarColumns

dx._groupby.DecomposableGroupbyAggregation
dx._groupby.GroupBy
dx._groupby.GroupbyAggregation
dx._groupby.SeriesGroupBy
dx._groupby.SingleAggregation

dx._reductions.Reduction
dx._reductions.Var

dx._shuffle._get_divisions
dx._shuffle.Shuffle
dx._shuffle.TaskShuffle

dx._util._convert_to_list
dx._util._raise_if_object_series
dx._util.is_scalar
```

</details>

The `dask_expr` imports are certainly more of a mess, because most of `dask_expr` is private, but Dask cuDF needs to use a lot of that private API to deal with differences between pandas and cuDF.

**Comment by phofl** [comment]
So dispatch and backends is exposed anyway, so this should be fine.

For everything else, I'd rather not guarantee that the path doesn't change. I added an api.py file in dask.dataframe that we can use to expose the stuff that other libraries need (underscore functions should probably be renamed to something without underscores and get docstrings if not already there).

Can you move the stuff over and put up a PR when you are adapting stuff over in dask-cudf?

**Comment by martindurant** [comment]
A deprecation cycle or advance warning would have been nice. dask-awkward and dask-histogram are currently broken. 
https://github.com/dask-contrib/dask-awkward/pull/566

**Comment by phofl** [comment]
The legacy implementation was deprecated in 2024.11.0
